weld/codegen/llvm2/cmp.rs

```diff
@@ -0,0 +1,323 @@
+                // if lt
+                let cond = gen_binop(builder, LessThan, left, right, ty)?;
+                LLVMBuildCondBr(builder, cond, done_block, on_geq_block);
+
+                LLVMPositionBuilderAtEnd(builder, on_geq_block);
+                // else if equal
+                // this is less likely to occur during sort so we don't check it if the lt branch passes
+                // TODO: special case booleans for Select
+                let eq = gen_binop(builder, Equal, left, right, ty)?;
+                let on_geq = LLVMBuildSelect(builder, eq, self.i32(0), self.i32(1), c_str!(""));
+                LLVMBuildBr(builder, done_block);
+                
+                // Finish block - set result.
+                LLVMPositionBuilderAtEnd(builder, done_block);
+                let result = LLVMBuildPhi(builder, ret_ty, c_str!(""));
+                let mut blocks = [entry_block, on_geq_block];
+                let mut values = [self.i32(-1), on_geq];
+                LLVMAddIncoming(result, values.as_mut_ptr(), blocks.as_mut_ptr(), values.len() as u32);
+
+                LLVMBuildRet(builder, result);
+            }
+            Struct(ref elems) => { // recursively apply cmp to elements
+                let mut result = self.i32(0);
+                for (i, elem) in elems.iter().enumerate() {
+                    let func = self.gen_cmp_fn(elem)?;
+                    let next_block = LLVMAppendBasicBlockInContext(self.context, function, c_str!(""));
+                    let done_block = LLVMAppendBasicBlockInContext(self.context, function, c_str!(""));
+
+                    let field_left = LLVMBuildStructGEP(builder, left, i as u32, c_str!(""));
+                    let field_right = LLVMBuildStructGEP(builder, right, i as u32, c_str!(""));
+
+                    let mut args = [field_left, field_right];
+                    // Compare struct field.
+                    let field_result = LLVMBuildCall(builder, func, args.as_mut_ptr(), args.len() as u32, c_str!(""));
+                    let cond = LLVMBuildICmp(builder, LLVMIntEQ, field_result, self.i32(0), c_str!(""));
+
+                    // Continue to next only if the field result returns 0 (fields are equal).
+                    LLVMBuildCondBr(builder, cond, next_block, done_block);
+                    
+                    // If field result is nonzero, return field result.
+                    LLVMPositionBuilderAtEnd(builder, done_block);
+                    LLVMBuildRet(builder, field_result);
+                    
+                    LLVMPositionBuilderAtEnd(builder, next_block);
+                }
+                
+                // All blocks equal, return 0.
+                LLVMBuildRet(builder, result); 
+            }
+            // Vectors comprised of unsigned chars or booleans can be compared for using `memcmp`.
+            Vector(ref elem) if elem.supports_memcmp() => {
+                let left_vector = self.load(builder, left)?;
```

Comment:
Sorry to nit, but can you actually change this to:

```rust
let left_data_ptr = LLVMBuildStructGEP(builder, left, vector::POINTER_INDEX, c_str!(""));
let left_data = self.load(builder, left_data_ptr)?;
let right_data_ptr = LLVMBuildStructGEP(builder, right, vector::POINTER_INDEX, c_str!(""));
let left_data = self.load(builder, right_data_ptr)?;

let left_size_ptr = LLVMBuildStructGEP(builder, left, vector::SIZE_INDEX, c_str!(""));
let right_size_ptr = LLVMBuildStructGEP(builder, right, vector::SIZE_INDEX, c_str!(""));
```

The reason for this is that adding alloca's in the middle of a function (as opposed to the very top) confuses the LLVM optimizer, and it sometimes can't figure out that it needs to move things to registers.